src/iop/hazeremoval.c

```diff
@@ -0,0 +1,973 @@
+  dt_bauhaus_slider_set(g->strength, p->strength);
+  dt_bauhaus_slider_set(g->distance, p->distance);
+
+  dt_pthread_mutex_lock(&g->lock);
+  g->distance_max = NAN;
+  g->A0[0] = NAN;
+  g->A0[1] = NAN;
+  g->A0[2] = NAN;
+  g->hash = 0;
+  dt_pthread_mutex_unlock(&g->lock);
+}
+
+void gui_init(dt_iop_module_t *self)
+{
+  self->gui_data = malloc(sizeof(dt_iop_hazeremoval_gui_data_t));
+  dt_iop_hazeremoval_gui_data_t *g = (dt_iop_hazeremoval_gui_data_t *)self->gui_data;
+  dt_iop_hazeremoval_params_t *p = (dt_iop_hazeremoval_params_t *)self->params;
+
+  dt_pthread_mutex_init(&g->lock, NULL);
+  g->distance_max = NAN;
+  g->A0[0] = NAN;
+  g->A0[1] = NAN;
+  g->A0[2] = NAN;
+  g->hash = 0;
+
+  self->widget = gtk_box_new(GTK_ORIENTATION_VERTICAL, DT_BAUHAUS_SPACE);
+
+  g->strength = dt_bauhaus_slider_new_with_range(self, -1, 1, 0.01, p->strength, 2);
+  dt_bauhaus_widget_set_label(g->strength, NULL, _("strength"));
+  gtk_box_pack_start(GTK_BOX(self->widget), GTK_WIDGET(g->strength), TRUE, TRUE, 0);
+  g_signal_connect(G_OBJECT(g->strength), "value-changed", G_CALLBACK(strength_callback), self);
+
+  g->distance = dt_bauhaus_slider_new_with_range(self, 0, 1, 0.005, p->distance, 3);
+  dt_bauhaus_widget_set_label(g->distance, NULL, _("distance"));
+  gtk_box_pack_start(GTK_BOX(self->widget), GTK_WIDGET(g->distance), TRUE, TRUE, 0);
+  g_signal_connect(G_OBJECT(g->distance), "value-changed", G_CALLBACK(distance_callback), self);
+}
+
+void gui_cleanup(dt_iop_module_t *self)
+{
+  dt_iop_hazeremoval_gui_data_t *g = (dt_iop_hazeremoval_gui_data_t *)self->gui_data;
+  dt_pthread_mutex_destroy(&g->lock);
+  free(self->gui_data);
+  self->gui_data = NULL;
+}
+
+//----------------------------------------------------------------------
+// module local functions and structures required by process function
+//----------------------------------------------------------------------
+
+typedef struct tile {
+  int left, right, lower, upper;
+} tile;
+
+typedef struct rgb_image {
+  rgb_pixel *data;
+  int width, height, stride;
+} rgb_image;
+
+static inline float * get_rgb_pixel(rgb_image img, size_t i)
+{
+  return ((float *)img.data) + i * img.stride;
+}
+
+typedef float gray_pixel;
+
+typedef struct gray_image {
+  gray_pixel *data;
+  int width, height;
+} gray_image;
+
+// allocate space for 1-component image of size width x height
+static inline gray_image new_gray_image(int width, int height)
+{
+  return (gray_image){ dt_alloc_align(64, sizeof(gray_pixel) * width * height), width, height };
+}
+
+// free space for 1-component image
+static inline void free_gray_image(gray_image *img_p)
+{
+  dt_free_align(img_p->data);
+  img_p->data = NULL;
+}
+
+// copy 1-component image img1 to img2
+static inline void copy_gray_image(gray_image img1, gray_image img2)
+{
+  memcpy(img2.data, img1.data, sizeof(gray_pixel) * img1.width * img1.height);
+}
+
+// minimum of two integers
+static inline int min_i(int a, int b)
```

Comment:
As darktable is written in C99, I replaced min/max functions for floats by fminf and fmaxf. However, I would prefer to use hand crafted inline functions for integers, rather than macros, in order to avoid double-evaluation of macro arguments.